### ISSUE TITLE
Don't bypass pytest-specific env vars to subprocesses

### ISSUE THREAD

**Issue by webknjaz** [pytest-dev/pytest#4359]
Hi, I'm currently trying to put Molecule test suite into Travis CI (https://github.com/ansible/molecule/pull/1541). It uses pytest.

(user story)
So everything was going well until I tried to split tests into chunks. I used "ephemeral" (dynamically generated) markers to split them into groups, so that using smth like `-m shard_1_of_3` selects one of those.
To avoid duplication of stuff across jobs in Travis CI config I've started using `PYTEST_ADDOPT='-m shard_1_of_3'`, which is when the fun began :)

So one of the tests started failing consistently (https://asciinema.org/a/5wWUdW1B9a7fvvFsMZTZq373J) and in mixed up with other side-effects, which made me think it's flaky. After I untangled everything, turned out, that functional test creates an env and runs molecule in that env.

What could possibly go wrong? It ran a subprocess with Molecule, which was running pytest (unsurprisingly, because Molecule is a testing framework) and that subprocess inherited all env vars including `PYTEST_ADDOPTS`¸ which of course made that pytest under subprocess (double subprocess now?) to not match anything and return a failuling return code.
What made things worse is tha Molecule seems to swallow output of pytest, which made it harder to debug. (I understand that it's a thing which has to be fixed within molecule, yes, but this issue about different thing).

So now I need a way to globally prevent such vars from affecting subprocesses, this is what I came up with:
```python
def pytest_sessionstart():
    """Make PYTEST_* env vars inaccessible to subprocesses."""                                      
    for var_name in list(os.environ):                                                               
        if var_name.startswith('PYTEST_'):                                                                 
            del os.environ[var_name]
```

The main reason of writing up this issue is basically to ask you whether it makes sense for pytest to have some built-in plugin/switch to enable such behavior out of the box. Or maybe mention in docs? What do you think?

P.S. Sharding implementation, not really related to the issue, but sharing for context:
```python
def pytest_collection_modifyitems(items):
￼    marker = pytest.config.getoption('-m')
￼    is_sharded = False
￼    shard_id = 0
    shards_num = 0
    if not marker.startswith('shard_'):
        return
    shard_id, _, shards_num = marker[6:].partition('_of_')
    if shards_num:
        shard_id = int(shard_id)
        shards_num = int(shards_num)
        is_sharded = True
    else:
        raise ValueError('shard_{}_of_{} marker is invalid')
    if not is_sharded:
        return
    if not (0 < shard_id <= shards_num):
        raise ValueError(
            'shard_id must be greater than 0 and not bigger than shards_num')
    for test_counter, item in enumerate(items):
        cur_shard_id = test_counter % shards_num + 1
        marker = getattr(pytest.mark, 'shard_{}_of_{}'.format(
            cur_shard_id,
            shards_num,
        ))
        item.add_marker(marker)
    del marker
    print('Running sharded test group #{} out of {}'.format(
￼        shard_id, shards_num))
```

**Comment by nicoddemus** [comment]
> So everything was going well until I tried to split tests into chunks. I used "ephemeral" (dynamically generated) markers to split them into groups

Just as a curiosity, I have thought about using this in the past, and the solution I thought was to use a custom `pytest_collection_modifyitems(items)` hook to select each Nth test in the items list based on the group index from an environment variable. This solution would not use the pytest mechanisms and would probably avoid the problem you are having altogether.

> So one of the tests started failing consistently... It ran a subprocess with Molecule, which was running pytest (unsurprisingly, because Molecule is a testing framework) and that subprocess inherited all env vars including PYTEST_ADDOPTS¸ which of course made that pytest under subprocess (double subprocess now?) to not match anything and return a failuling return code.

The subprocess invocation happens inside the test code? If so I would write a fixture and explicitly use that fixture in that test:

```python
@pytest.fixture
def reset_pytest_vars(monkeypatch):
    """Make PYTEST_* env vars inaccessible to subprocesses."""
    for var_name in os.environ:
        if var_name.startswith('PYTEST_'):
            monkeypatch.delenv(var_name, raising=False)
```

This solution is reasonable if this kind of test (which invokes a Molecule subprocess) is rare, but you can also make it autouse if you think more tests like this are likely to be added in the future.

> The main of writing up this issue is basically to ask you whether it makes sense for pytest to have some built-in plugin/switch to enable such behavior out of the box. Or maybe mention in docs? What do you think?

Thanks for taking the time to write, we appreciate it. In my opinion though this case seems really exotic, so I don't think we should add anything to pytest (docs or otherwise) because it seems would not really add much to users.

**Comment by webknjaz** [comment]
> This solution would not use the pytest mechanisms and would probably avoid the problem you are having altogether.

It's an interesting idea! Thanks for sharing! I might be changing my solution with this suggestion in mind :)

My primary goal was to "just make it work" because we (@ansible) officially adopted Molecule recently and enabled CI is a must. So I thought I'd fix things minimally and go for improvements later because blocking contributors from merging PRs just because of missing CI config is not fun :)

First thing I tried was googling: I truly hoped to find out some suitable prior art plugin and just use it. I've found https://pypi.org/project/pytest-test-groups, but it turned out to be unmaintained and has a critical bug in group selection logic :(
I also know that Circle CI has their CLI client, which does group selection on its own, but it's not suitable for a number of reasons (including, but not limited to the fact we use Travis CI).

So I ended up trying to implement something on my own, which is not perfect (`-m` check is extremely silly), but works.


> I would write a fixture [...] make it autouse

Thanks for this too! There are multiple things invoking subprocesses in tests, that's why I tried to hook into the process as soon as possible. I'll try your suggestion, it looks very reasonable :)


> this case seems really exotic

Yeah, that's what I thought, but I also wanted to post it in some googlable place like this issue :)
